do/csi/driver/controller.go

```diff
@@ -3,48 +3,345 @@ package driver
-// accesible by anymore.
-func (d *Driver) DeleteVolume(context.Context, *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
-	return nil, errors.New("not implemented")
+// DeleteVolume deletes the given volume. The function is idempotent.
+func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "DeleteVolume Volume ID must be provided")
+	}
+
+	_, err := d.doClient.Storage.DeleteVolume(ctx, req.VolumeId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &csi.DeleteVolumeResponse{}, nil
 }
 
-// ControllerPublishVolume ...
-func (d *Driver) ControllerPublishVolume(context.Context, *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
-	return nil, errors.New("not implemented")
+// ControllerPublishVolume attaches the given volume to the node
+func (d *Driver) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "ControllerPublishVolume Volume ID must be provided")
+	}
+
+	if req.NodeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "ControllerPublishVolume Node ID must be provided")
+	}
+
+	if req.VolumeCapability == nil {
+		return nil, status.Error(codes.InvalidArgument, "ControllerPublishVolume Volume capability must be provided")
+	}
+
+	dropletID, err := strconv.Atoi(req.NodeId)
+	if err != nil {
+		return nil, fmt.Errorf("malformed nodeId %q detected: %s", req.NodeId, err)
+	}
+
+	// TODO(arslan): wait volume to attach
+	_, resp, err := d.doClient.StorageActions.Attach(ctx, req.VolumeId, dropletID)
+	if err != nil {
+		// don't do anything if attached
+		if resp.StatusCode == http.StatusUnprocessableEntity || strings.Contains(err.Error(), "This volume is already attached") {
+			return &csi.ControllerPublishVolumeResponse{}, nil
+		}
+
+		return nil, err
+	}
+
+	return &csi.ControllerPublishVolumeResponse{}, nil
 }
 
-// ControllerUnpublishVolume ...
-func (d *Driver) ControllerUnpublishVolume(context.Context, *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
-	return nil, errors.New("not implemented")
+// ControllerUnpublishVolume deattaches the given volume from the node
+func (d *Driver) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "ControllerPublishVolume Volume ID must be provided")
+	}
+
+	dropletID, err := strconv.Atoi(req.NodeId)
+	if err != nil {
+		return nil, fmt.Errorf("malformed nodeId %q detected: %s", req.NodeId, err)
+	}
+
+	// TODO(arslan): wait volume to deattach
+	_, resp, err := d.doClient.StorageActions.DetachByDropletID(ctx, req.NodeId, dropletID)
```

Comment:
I believe this needs a `s/req.NodeId/req.VolumeId`, should be:
`_, resp, err := d.doClient.StorageActions.DetachByDropletID(ctx, req.VolumeId, dropletID)`